AstSemantics.md

```diff
@@ -111,6 +112,9 @@ Integer loads can specify a *storage size* which is smaller than the result type
 well as a signedness which determines whether the bytes are sign- or zero-
 extended into the result type.
 
+Each `load` and `store` specifies the [linear memory index](Modules.md#index-spaces)
+to operate on with an unsigned integer immediate.
+
```

Comment:
> In addition to "file system impl", "asm.js globals" is another use cases where you have a simple, static separation.

Yes, it would be nice for that. But we've been ok without it, and it's temporary anyhow.

> Another cool use case could be a *San tool that wants to maintain a whole shadow data structure to reflect memory; a separate memory is exactly what you want; with a single memory, you'd have to prematurely fix the heap size to split "shadow" from "real" and then "real" can also corrupt "shadow".

Yes, but that use case is well handled by a GC ref to a buffer, as the addition of bounds checks on the shadow data is not significant when doing *Sanning. There are also some *San use cases that would need a GC ref anyhow, consider a thread sanitizer that creates a buffer per thread, the number of threads may not be known at compile time.

> Also, if you take a modular-web-app POV (instead of a monolithic-C++-app POV), there's not just 1 big heap; you have lots of modules in lots of heaps. You don't want to force them all into the same heap (where they may clobber each other) or force them to do unnecessary copying if they form a data pipeline. GC-Memory references handles the general case, but imports/exports would allow the pipeline to be declaratively composed using the plain old module system.

First, if we're talking non-C++, would heaps even be used? Something like Java or C# might be compiled to use GC buffers anyhow. Or, if you meant C++, how would that work exactly - a pointer would only be able to refer to one of those heaps? Would the modules not communicate through shared pointers? (If so I don't think they need this feature?)

Second, in your proposal, Memories arrive at import time, i.e., when a module is instantiated? Then two modules can't refer to each other's Memories, can they? A further limitation is that this wouldn't be able to handle dlopen() style linking, that happens at runtime. (A GC ref to buffers, on the other hand, doesn't have any limitation and can handle arbitrary linking.)

Overall, I don't mean to pick at your specific examples. My general point is that, as you said, a GC ref to a buffer is a generalization of this, and we'll have it anyhow, so multiple Memories adds complexity to the spec and to VMs that is not strictly necessary. Perhaps an important use case that needs to avoid the sandboxing overhead of GC buffers could justify it, but the use cases so far seem kind of unusual. Also, we don't have data on that GC ref overhead either yet, so this seems premature.

> (2) obvious symmetry in design given that we have multi function/table export

I would argue that import/export of a Table, and having that Table be modifiable, is much more important than having multiple imported/exported Tables. A single Table shared between modules that can be changed is all dynamic linking needs to get good performance.

With that said, I agree that the symmetry argument is a good one.